### PR TITLE
Makes scaling optional to keep original image dimension

### PR DESCRIPTION
**Pull request by repraze**
Open for comments on overall solution.

 - Adds a scaling toggle to skip the fill/fit/stretch mode.
 - Renamed `scaling` to `scalingMode` and reused `scaling` to contain Boolean value.
 - Makes scaling enabled by default. Could potentially change that. (Still needs proper testing on a fresh start with no data)
 
This solution allows the OS to do the sizing and enables proper parallax on pixel OS.

| Disabled | Enabled |
| :-: | :-: |
|  |  |

Considerations
 - An alternative implementation would be to add another ScalingConstants with `None` or `Default` to let the OS take over the scaling. The radio input seems crowded for a fourth option though. I thought the additional checkbox was in line with other toggleable settings.
 - The new Resize string is not properly translated. I wasn't sure about the process here.
 - Not sure if renaming scaling to scalingMode might have non-backward compatible issue with reading the saved settings. If so, we could keep the string binding the same for backward compatibility.
 
relates to https://github.com/Anthonyy232/Paperize/issues/92

### CODE REVIEW

File: app/src/main/java/com/anthonyla/paperize/core/SettingsConstants.kt [hunk 1]
```diff
@@ -28,6 +28,7 @@ object SettingsConstants {
     const val LOCK_DARKEN_PERCENTAGE = "darken_darken_percentage"
     const val DARKEN = "darken"
     const val WALLPAPER_SCALING = "wallpaper_scaling"
+    const val WALLPAPER_SCALING_MODE = "wallpaper_scaling_mode"
```
Comment (repraze) [inline app/src/main/java/com/anthonyla/paperize/core/SettingsConstants.kt:31]: wallpaper_scaling now contains the enablement and wallpaper_scaling_mode the value

File: app/src/main/java/com/anthonyla/paperize/core/WallpaperUtil.kt [hunk 2]
```diff
@@ -345,18 +345,21 @@ fun processBitmap(
...
 ): Bitmap? {
     try {
         var processedBitmap = source
 
         // Apply wallpaper scaling effects
-        processedBitmap = when (scaling) {
-            ScalingConstants.FILL -> fillBitmap(processedBitmap, device.widthPixels, device.heightPixels)
-            ScalingConstants.FIT -> fitBitmap(processedBitmap, device.widthPixels, device.heightPixels)
-            ScalingConstants.STRETCH -> stretchBitmap(processedBitmap, device.widthPixels, device.heightPixels)
+        if (scaling) {
```
Comment (repraze) [inline app/src/main/java/com/anthonyla/paperize/core/WallpaperUtil.kt:357]: skip when scaling disabled

File: app/src/main/java/com/anthonyla/paperize/feature/wallpaper/presentation/wallpaper_screen/components/WallpaperPreviewAndScale.kt [hunk 3]
```diff
@@ -82,88 +85,106 @@ fun WallpaperPreviewAndScale(
...
+                    modifier = Modifier.padding(16.dp),
+                    fontWeight = FontWeight.W500
+                )
+                Switch(
+                    checked = scaling,
+                    onCheckedChange = onScalingChange
+                )
+            }
+
+            if (scaling) {
```
Comment (repraze) [inline app/src/main/java/com/anthonyla/paperize/feature/wallpaper/presentation/wallpaper_screen/components/WallpaperPreviewAndScale.kt:100]: hide resize options when scaling disabled

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -162,4 +162,5 @@
     <string name="enable_wallpaper_changer">Enable Wallpaper Changer</string>
     <string name="unselected_album">Unselected album</string>
     <string name="lock">Lock</string>
+    <string name="resize">Resize</string>
```
Comment (repraze) [inline app/src/main/res/values/strings.xml:165]: needs translation - not sure of the process.

**Comment by repraze** [comment]
An observation:

pixel OS does resize the image to allow parallax - it takes into account the number of home screens and will do a 'fill' with overflow on the right.

The first screen is centered on the image and you can swipe right to reveal more.

By setting a raw image without resize, parallax works, but it will be left aligned. The first screen shows the left of the image and you can swipe right to reveal more.

I think it might be hard to account for all manufacturer implementations of the home screen/wallpaper positioning. This solution is good enough, and the user can always crop/position the images to achieve the same effect.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (Anthonyy232) [reply app/src/main/res/values/strings.xml:165]: Can the string be changed to <string name="apply_image_scaling">Apply Image Scaling</string>

File: app/src/main/java/com/anthonyla/paperize/core/SettingsConstants.kt (hunk 1 above)
Comment (Anthonyy232) [reply app/src/main/java/com/anthonyla/paperize/core/SettingsConstants.kt:31]: In SettingsViewModle Reset, WALLPAPER_SCALING needs to be changed from deleteString to deleteBoolean, and then add a deleteString for WALLPAPER_SCALING_MODE.

**Comment by Anthonyy232** [comment]
Yeah, unfortunately I can't even check the amount of home screens -- even within the WallpaperManager class, many are restricted as system APIs which is a little frustrating as the feature set is very neutered.

**Comment by repraze** [comment]
I'll play with it some more. I'll make this a draft until it's in a ready state.

I have a few ideas to make resizing / scaling more powerful. But I want to make sure the UI stays nice and friendly.

**Comment by Anthonyy232** [comment]
Implemented this in commit https://github.com/Anthonyy232/Paperize/commit/9f67243f0dd5d2ea350408dea2bad2bdaa1ea082 by adding another scaling option. Will be closing this, feel free to create another PR if you figure this out!

image

**Pull request closed without merging**